protocols/ML01.md

```diff
@@ -0,0 +1,315 @@
+# ML01: Machine learning basic specification
+**THIS IS A DRAFT SPECFICATION AND IS NOT FINAL**
+
+| name                       | ML01               |
+|----------------------------|--------------------|
+| version                    | 0.1.0              |
+| status                     | proposal           |
+| author                     | Ryan Wilson (rw86) |
+| serving component(s)       | Machine Learning   |
+| consuming component(s)     | HCI                |
+| basic spec                 | yes                |
+| can be required by servers | no                 |
+| can be required by clients | HCI                |
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+
+## Changelog
+### Version 0.1
+
+- First draft
+
+# Foreword: Acknowledgements
+
+This specification is loosely based on ideas presented in an endpoint outline "ml_endpoints.md" written by Daphne Bogosian. There are some major architectural differences, outlined below:
+
+- The principal "backend should be responsible for all matters of persistance" agreed during protocol meetings has been incorporated
+  - In particular results are returned to the frontend as files via backend (this was also agreeed)
+- The principal "HCI should talk directly to backend" has been incorporated.
+  - In particular there is no BE <-> ML communication.
+- The API has been generalised to not specifically reference training/models and instead refence the more general "jobs":
+  - Gives more flexibility to the implementation for data processing tasks that do not fit this pattern.
+  - Removes burden on client.
+  - Jobs referencing models are still posibile (see implementation notes)
+- Response format made consistent with BE01.
+- Spec now only uses GET and POST, as these are the request types with most widely available browser support.
```

Comment:
I thought mostly all browsers supported the other request types as well?

Reply:
Was talking with Jordan about this as well  - I'm still not entirely clear about the client support for RESTFUL methods but this may have been too conservative a design choice - no issue with changing it.

Reply:
So if you do change this so that requests are just forwarded to ML, there should be no issue making ML requests other than GET/POST since it won't be through a browser?

Reply:
But the browser would still make a request to BE and (unless the BE modifies the request method before forwarding) there is still a question of client support.

Reply:
Yeah, you're right. I think most browsers should support other requests though, and it would make the API a lot nicer to use some of the other ones.

Reply:
Jordan's issue: https://github.com/rtw43/cs3099jh/issues/4 (this was specifically backend, but relevant)